dev/mailing#95 Only track <a> urls in Flexmailer for HTML emails

larssandergreen opened the pull request:
[Issue here.](https://lab.civicrm.org/dev/mail/-/issues/95)

Before
----------------------------------------
Link tracking was applied to all URLs inside tags, including tags like &lt;link&gt;.

After
----------------------------------------
Link tracking is applied only to URLs inside &lt;a&gt; tags. Regex simply looks for &lt;a anything href=URL&gt; instead of &lt;anything href=URL&gt; (edited to add "anything" for clarity).

MikeyMJCO commented:
The `href` does not have to immediately follow the `<a` and `<a title="Title" href="link">Text</a>` is perfectly valid HTML which this should account for.

larssandergreen commented:
@artfulrobot If you use A in a tag and save the mailing, it is converted to lowercase (same for HREF, which wouldn't have worked before now if not). But good idea to add i here — belt and braces.

That does take the PR up to four characters though.

larssandergreen commented:
@MikeyMJCO Sorry, I oversimplified in my explanation and have now clarified. All I've changed is where the regex formerly looked for < to start matching the pattern, now it looks for <a. (and I've now added an i to make this case insensitive). But just to be sure, I just tested with <code><a title="Title" href="link">Text</a></code>.

eileenmcnaughton commented:
@artfulrobot @MikeyMJCO do either of you have a feeling what the status is on this one?

MikeyMJCO commented:
Approved from me - small but valuable change. Code is clear and a few things clarified by the author in-transit!

eileenmcnaughton commented:
I've added merge-ready just in case @artfulrobot wants to comment before we merge based on your review @MikeyMJCO

totten commented:
* Agree tracking `<link rel="stylesheet" href="styles.css">` would be superfluous. (In the current state of affairs, you'd expect stats on that should be similar to the pixel-tracking... so it's redundant.) 
* Dunno if email documents ever use `<base>` tags... tracking relative URLs is probably a sensible thing, but it's ill-defined in the current schema. Adding the tracking-code to `<base>` is probably harmful at the moment.
* [`<area href>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/map) is more of a  bona-fide/clickable hyperlink. OTOH, `<area href>` probably died off ten years ago? It seems more correct to track it, but I wouldn't block other fixes on that.

totten commented:
Shoulda read the regex... `;(\<a[^>]*href *= *")([^">]+)(");i',` probably matches both `<a ... href>` and `<area ... href>`. 🙃

eileenmcnaughton commented:
@totten is that a thumbs up on merging?

totten commented:
Yeah, concept+approach seem fine. Haven't `r-run`. However, the test coverage in `HtmlClickTrackerTest`, `ClickTrackerTest`, and `FlexMailerSystemTest::testUrlTracking()` all seem relevant.

larssandergreen commented:
Thanks all!
